### PR TITLE
change score breakdown from array to map

### PR DESCRIPTION
**Pull request by CascadingRadium**
looking at https://issues.couchbase.com/browse/MB-59821
it seems that having score breakdown as a slice would be inoptimal in other use cases of disjunction searchers such as in Prefix and Wildcard queries. 
Here the number of searchers involved in a single disjunction searcher may be very large and can affect cpu execution time.
This approach changes the datatype of scorebreakdown to a map

### CODE REVIEW

File: search_knn.go [hunk 1]
```diff
@@ -225,7 +225,7 @@ func adjustRequestSizeForKNN(req *SearchRequest, numIndexPartitions int) int {
 
 // heap impl
 type scoreHeap struct {
-	scoreBreakdown []*[]float64
+	scoreBreakdown []*map[int]float64
```
Comment (abhinavdangeti) [inline search_knn.go:228]: Does this need to be an array of maps?

Also the map does not need to be a pointer, just `[]map[int]float64` should do.

File: search/search.go [hunk 2]
```diff
@@ -181,7 +181,7 @@ type DocumentMatch struct {
 	// the same as the number of sub-queries in the query.
 	// the order of the scores will match the order of the sub-queries
 	// in the query.
-	ScoreBreakdown []float64 `json:"score_breakdown,omitempty"`
+	ScoreBreakdown map[int]float64 `json:"score_breakdown,omitempty"`
```
Comment (abhinavdangeti) [inline search/search.go:184]: Update the commentary above on what the key-values in this map represent.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

**Review by abhinavdangeti — COMMENTED** [review]
@CascadingRadium I thought we spoke of determining the score breakdown and shipping it _only_ when a knn request comes in ? If you are doing that already, I can't quite seem to spot it - would you point me to it.

**Comment by CascadingRadium** [comment]
yea but i wanted to check if the regressions go away with just the optimization in this PR. if the results arent so good, i will add that check, in a different PR.

**Comment by abhinavdangeti** [comment]
Per conversation w/ @CascadingRadium - new patch coming here addressing above comments.

File: search/query/conjunction.go [hunk 3]
```diff
@@ -26,9 +26,10 @@ import (
 )
 
 type ConjunctionQuery struct {
-	Conjuncts       []Query `json:"conjuncts"`
-	BoostVal        *Boost  `json:"boost,omitempty"`
-	queryStringMode bool
+	Conjuncts              []Query `json:"conjuncts"`
+	BoostVal               *Boost  `json:"boost,omitempty"`
+	RetrieveScoreBreakdown bool    `json:"retrieve_score_breakdown,omitempty"`
```
Comment (abhinavdangeti) [inline search/query/conjunction.go:31]: Yo, I don't like the feel of this at all.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]
Our conversation was this -
* Ok to do a score breakdown when a compound query is involved.
* I wanted us to put in an optimization where we do optimize a compound query with kNN to simply a kNN when possible -
    * `match_none` OR single `kNN`
    * `match_all` AND single `kNN`
* It is ok to do a score breakdown for multi kNN

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**